app/code/core/Mage/Usa/Model/Shipping/Carrier/Ups.php

```diff
@@ -1793,4 +1697,398 @@ protected function _getDeliveryConfirmationLevel($countyDest = null)
+                        && $this->getConfigData('shipper_number')
+                        && !empty($negotiatedArr);
+
+                    $this->processShippingRestRateForItem(
+                        $shipElement,
+                        $allowedMethods,
+                        $allowedCurrencies,
+                        $costArr,
+                        $priceArr,
+                        $negotiatedActive
+                    );
+                }
+            } else {
+                $errorTitle = $rateResponseData['RateResponse']['Response']['ResponseStatus']['Description'] ?? '';
+                $error = Mage::getModel('shipping/rate_result_error');
+                $error->setCarrier('ups');
+                $error->setCarrierTitle($this->getConfigData('title'));
+                $error->setErrorMessage($this->getConfigData('specificerrmsg'));
+            }
+        }
+
+        return $this->setRatePriceData($priceArr, $costArr, $errorTitle);
+    }
+
+    private function setRatePriceData(array $priceArr, array $costArr, string $errorTitle): Mage_Shipping_Model_Rate_Result
+    {
+        $result = Mage::getModel('shipping/rate_result');
+
+        if (empty($priceArr)) {
+            $error = Mage::getModel('shipping/rate_result_error');
+            $error->setCarrier('ups');
+            $error->setCarrierTitle($this->getConfigData('title'));
+            if ($this->getConfigData('specificerrmsg') !== '') {
+                $errorTitle = $this->getConfigData('specificerrmsg');
+            }
+            $error->setErrorMessage($errorTitle);
+            $result->append($error);
+            return $result;
+        }
+
+        foreach ($priceArr as $method => $price) {
+            $shipmentDescription = $this->getShipmentByCode($method);
+            if (!strlen($shipmentDescription)) {
+                continue;
+            }
+
+            $rate = Mage::getModel('shipping/rate_result_method');
+            $rate->setCarrier('ups');
+            $rate->setCarrierTitle($this->getConfigData('title'));
+            $rate->setMethod($method);
+            $rate->setMethodTitle($shipmentDescription);
+            $rate->setCost($costArr[$method]);
+            $rate->setPrice($price);
+            $result->append($rate);
+        }
+
+        return $result;
+    }
+
+    /**
+     * Processing rate for ship element
+     *
+     * @param array $shipElement
+     * @param array $allowedMethods
+     * @param array $allowedCurrencies
+     * @param array $costArr
+     * @param array $priceArr
+     * @param bool $negotiatedActive
+     */
+    private function processShippingRestRateForItem(
+        array $shipElement,
+        array $allowedMethods,
+        array $allowedCurrencies,
+        array &$costArr,
+        array &$priceArr,
+        bool $negotiatedActive
+    ): void {
+        $code = $shipElement['Service']['Code'] ?? '';
+        if (true or in_array($code, $allowedMethods)) { // TODO CHECK FAILS
```

Comment:
this fails because $allowedMethods is populated wrong, and it's wrong because in ups.phtml the changeOriginShipment is not run for some reason, which changes the available methods depending on the origin

Reply:
Can you check again, I fixed a JS error that may have been causing this.

Reply:
great! seems fixed!